types/types.go

```diff
@@ -65,18 +68,102 @@ type Result struct {
 	Log []LogAttribute `json:"log"`
 }
 
+// TODO: get this proper
+type ApiError struct {
+	Base64Err      *SourceErr     `json:"base64_err,omitempty"`
+	ContractErr    *MsgErr        `json:"contract_err,omitempty"`
+	DynContractErr *MsgErr        `json:"dyn_contract_err,omitempty"`
+	NotFound       *NotFoundErr   `json:"not_found,omitempty"`
+	NullPointer    *struct{}      `json:"null_pointer,omitempty"`
+	ParseErr       *JSONErr       `json:"parse_err,omitempty"`
+	SerializeErr   *JSONErr       `json:"serialize_err,omitempty"`
+	Unauthorized   *struct{}      `json:"unauthorized,omitempty"`
+	UnderflowErr   *UnderflowErr  `json:"underflow_err,omitempty"`
+	Utf8Err        *SourceErr     `json:"utf8_err,omitempty"`
+	Utf8StringErr  *SourceErr     `json:"utf8_string_err,omitempty"`
+	ValidationErr  *ValidationErr `json:"validation_err,omitempty"`
+}
```

Comment:
I'm not that well versed in Go, but is it maybe better to define this as something akin to a pointer to a trait object in rust? that way the layout of this type won't be so big.

Reply:
Yes, I can try to use a (private) interface and then separate types/structs for each of these cases, and use a type-switch. This guarantees 0 or 1 elements (not 0 to N) and will be a smaller struct, but actually a fair bit more code for the structs.

Reply:
That probably is more idiomatic Go as well

Reply:
I started doing this for errors and it is a lot of code. And they are being overhauled right now.. 

I will update messages to work like this, and then update errors once https://github.com/CosmWasm/cosmwasm/issues/271 is merged and we update go-cosmwasm to that commit

Reply:
Oof... I did this for `CosmosMsg`, then I realized something...
`interface{}` types can be json serialized, but not deserialized (missing info on how), so I will need more custom code there.

I was trying the approach used by protobuf, but they use codegen and determine the bytes and cast it to a type manually (in generated code).

I will stick with the current (ugly) solution. But at least I know why I did this

Reply:
Sounds reasonable. The overhead here isn't too terrible, and it's definitely fixable in the future since there's zero effect on exposed interfaces afaik.
How/Where do these types propagate to the cosmos sdk and eventually users of the CLI or client libraries?